### PR TITLE
feat(cli): update to TypeScript 4.1

### PR DESCRIPTION
**Pull request by kitsonk**
This PR updates Deno to TypeScript 4.1 beta in advance of TypeScript 4.1 being released to give an early view into any issues.

TypeScript 4.1 tightens up `Promise` definitions and there are quite a few instances in the code base where the type checking is failing due to this.

### CODE REVIEW

**Comment by Skillz4Killz** [comment]
Hey 👋 

I was wondering what the plans are for `--noUncheckedIndexedAccess`? This was added in TS 4.1 and is probably one of the greatest additions to TS since forever and I did not see any mention of it. Recently, Deno made an update which enabled stricter configurations in tsconfig by default for imports. I believe this should also apply to `noUncheckedIndexedAccess`. It prevents so many issues/bugs because without it the types provided by TS are just plain wrong. One of the best parts I love about Deno is that you can remove all the extra files like `tsconfig`, `eslintrc`, `prettier` etc... Having to create a tsconfig just to have this enabled is kinda silly. I believe Deno by default should always encourage safety/strictness!

Thank You

**Comment by kitsonk** [comment]
@Skillz4Killz my personal opinion is that we don't enable it by default, as mentioned in the PR for it, the TypeScript core team says that they won't make it part of `--strict` mode (yet) and they don't think it is a sensible default (unlike the rest of the _strict_ options):

> - We do not intend to make this part of the `--strict` family with its current behavorial limitations, so anything starting with strict is a no-go
>   - If, in the future, we find a way to make this palatable (i.e. can be turned on with nearly zero false positives), we can make it be an alias for a new `--strictIndexSignatures` setting
> - We don't really think this is a great mode to be in by default, so the name should connote "Here be dragons" in terms of ergonomics

While I agree it is a really useful feature for very strongly typed code, I don't think we can unleash it as a default on everyone.  People can of course still enable it via use of the `--config` flag.  We may want to experiment with `std` though, just as we ensured that `std` was _strict_ safe before we turned it on by default.

File: std/node/_fs/_fs_writeFile_test.ts
Comment (bartlomieju) [inline std/node/_fs/_fs_writeFile_test.ts:104]: Is type parameter now mandatory for `Promise`?

**Review by bartlomieju — APPROVED** [review]
LGTM

File: std/node/_fs/_fs_writeFile_test.ts
Comment (kitsonk) [reply std/node/_fs/_fs_writeFile_test.ts:104]: If it cannot be inferred, yes. https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#resolves-parameters-are-no-longer-optional-in-promises. A good type safety thing that avoids a lot of errors of resolving promises incorrectly.

**Comment by bartlomieju** [comment]
> TypeScript 4.1 tightens up `Promise` definitions and there are quite a few instances in the code base where the type checking is failing due to this.

Maybe we should hold off this update till Deno `1.6.0` then?

**Pull request merged**
